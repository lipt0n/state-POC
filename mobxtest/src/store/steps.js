import { observable, action,flow } from "mobx"
import {getSteps,runStep,changeConfig} from '../api'

export default class StepsStore {
    constructor(rootStore) {
        this.rootStore = rootStore
      }
    fetchInterval = null;  
    @observable steps = []
    @observable state = "pending" // "pending" / "done" / "error"
    @action watchStart = ()=>{
        this.fetchInterval = setInterval(this.fetchSteps,10000)
    }
    @action watchStop = ()=>{
        clearInterval(this.fetchInterval)
    }
    fetchSteps = flow(function* () { 
        this.state = "pending"
        try {
            const {activeProject, activeWorkflow} = this.rootStore.projectsStore
            let steps = yield fetch(...getSteps(activeProject, activeWorkflow))
            steps = yield steps.json()
            this.state = "done"
            this.steps = steps
        } catch (error) {
            this.state = "error"
        }
    }).bind(this)
    runStep = flow(function*(stepId){
        const {activeProject, activeWorkflow} = this.rootStore.projectsStore
        try {
            const run = yield fetch(...runStep(activeProject, activeWorkflow,stepId))
            this.fetchSteps()
        } catch (error) {
            this.state = "error"
        }
    }).bind(this)
    changeStepConfig = flow(function*(stepId,config){
        const {activeProject, activeWorkflow} = this.rootStore.projectsStore
        try {
            const run = yield fetch(...changeConfig(activeProject, activeWorkflow,stepId,config))
            this.fetchSteps()
        } catch (error) {
            this.state = "error"
        }
    }).bind(this)
}