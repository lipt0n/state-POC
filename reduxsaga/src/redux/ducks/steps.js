import {showError} from './error'
import {getSteps as getStepsApi, runStep as runStepApi, changeConfig as changeConfigApi} from '../../api'
import {getActiveProject} from './projects'
import {getActiveWorkflow} from './workflows'
import { select, put, takeLatest,delay,fork }  from 'redux-saga/effects'

export const TYPES = {
    get:'STEPS:GET',
    set:'STEPS:SET',
    change:'STEPS:CHANGE',
    run:'STEPS:RUN'
}

export const getSteps = () => ({
    type:TYPES.get
})
export const setSteps= (steps) => ({
    type:TYPES.set,
    payload:steps
})
export const runStep= (stepId) => ({
    type:TYPES.run,
    payload:{
        stepId
    }
})
export const changeStepConfig= (stepId,config) => ({
    type:TYPES.change,
    payload:{
        stepId,
        config
    }
})
export const stepsReducer = (state=[], action) => {
    if(action.type === TYPES.set) return action.payload
    return state
}
export function* sagaFetch(activeProject,activeWorkflow){
    let steps = yield fetch(...getStepsApi(activeProject,activeWorkflow))
    steps = yield steps.json()
    yield put(setSteps(steps))
}
export function* sagaInterval(){
try {
    const activeProject = yield select(getActiveProject)
    const activeWorkflow = yield select(getActiveWorkflow)
    while(true){
        yield fork(sagaFetch,activeProject,activeWorkflow)
        yield delay(10000)
    }
  } catch (error) {
    yield put(showError(error))
  }
}
export function* sagaRun(action){
    try {
        const activeProject = yield select(getActiveProject)
        const activeWorkflow = yield select(getActiveWorkflow)
        yield fetch(...runStepApi(activeProject,activeWorkflow,action.payload.stepId))
        yield fork(sagaFetch,activeProject,activeWorkflow)

      } catch (error) {
        yield put(showError(error))
      }
}
export function* sagaChangeConfig(action){
    try {
        const activeProject = yield select(getActiveProject)
        const activeWorkflow = yield select(getActiveWorkflow)
        yield fetch(...changeConfigApi(activeProject,activeWorkflow,action.payload.stepId, action.payload.config))
        yield fork(sagaFetch,activeProject,activeWorkflow)

        } catch (error) {
        yield put(showError(error))
        }
}
export function* initStepsSagas() {
  yield takeLatest(TYPES.get, sagaInterval)
  yield takeLatest(TYPES.run, sagaRun)
  yield takeLatest(TYPES.change, sagaChangeConfig)

}
export const selectSteps = state => state.steps