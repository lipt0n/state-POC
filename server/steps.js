export default [
    {
      "id":0,
      "operator_id": "load_int",
      "last_executed_at": undefined,
      "status": "new",
      "configuration": [
        {
          "value":1
        }
      ],
      "inputs": [
       
      ]
    },
    {
      "id":1,
      "operator_id": "increment_int",
      "last_executed_at": undefined,
      "status": "new",
      "configuration": [
        {
          "increment": true
        }
      ],
      "inputs": [
        {
            "other_step_id": 0,
            "other_step_type_output_id": "value",
            "step_type_input_id": "value"
          }
      ]
    },
    { "id":2,
      "operator_id": "increment_int",
      "last_executed_at": undefined,
      "status": "new",
      "configuration": [
        {
          "increment": false
        }
      ],
      "inputs": [
        {
            "other_step_id": 1,
            "other_step_type_output_id": "value",
            "step_type_input_id": "value"
          }
      ]
    },
    { "id":3,
      "operator_id": "increment_int",
      "last_executed_at": undefined,
      "status": "new",
      "configuration": [
        {
          "increment": true
        }
      ],
      "inputs": [
        {
            "other_step_id": 2,
            "other_step_type_output_id": "value",
            "step_type_input_id": "value"
          }
      ]
    },
  ]