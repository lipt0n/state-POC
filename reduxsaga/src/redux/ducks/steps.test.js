
import * as steps from './steps'
import { select, put, takeLatest,delay,fork }  from 'redux-saga/effects'
import { expectSaga } from 'redux-saga-test-plan'
beforeEach(() => {
    fetch.resetMocks()
  })

const stepsObj = [{"id":0,"operator_id":"load_int","last_executed_at":"2019-03-25T21:50:58.643Z","status":"up_to_date","configuration":[{"value":1}],"inputs":[],"outputs":[{"value":1}]},{"id":1,"operator_id":"increment_int","last_executed_at":"2019-03-25T22:00:41.906Z","status":"up_to_date","configuration":[{"increment":0}],"inputs":[{"other_step_id":0,"other_step_type_output_id":"value","step_type_input_id":"value"}],"outputs":[{"value":1}]},{"id":2,"operator_id":"increment_int","last_executed_at":"2019-03-25T22:01:11.145Z","status":"up_to_date","configuration":[{"increment":1}],"inputs":[{"other_step_id":1,"other_step_type_output_id":"value","step_type_input_id":"value"}],"outputs":[{"value":2}]},{"id":3,"operator_id":"increment_int","last_executed_at":"2019-03-25T22:01:13.904Z","status":"up_to_date","configuration":[{"increment":0}],"inputs":[{"other_step_id":2,"other_step_type_output_id":"value","step_type_input_id":"value"}],"outputs":[{"value":2}]}]

test('getSteps action creator',()=>{
    expect(steps.getSteps()).toMatchSnapshot()
})
test('setSteps action creator',()=>{
    expect(steps.setSteps(stepsObj)).toMatchSnapshot()
})
test('runStep action creator',()=>{
    expect(steps.runStep(1)).toMatchSnapshot()
})
test('changeStepConfig action creator',()=>{
    expect(steps.changeStepConfig(1,{"increment":0})).toMatchSnapshot()
})
test('stepsReducer',()=>{
    expect(steps.stepsReducer(undefined, {type:"something else"}).length).toEqual(0)
    expect(steps.stepsReducer(undefined,steps.setSteps(stepsObj))).toMatchSnapshot()
})
// normal testing using .next .next .next 
test('sagaFetch*',()=>{
    const sagaFetch = steps.sagaFetch(1,1)
    const getJsonMock = jest.fn()
    getJsonMock.mockReturnValue(stepsObj)
    sagaFetch.next() // yield fetch(...getStepsApi(activeProject,actinpm startveWorkflow))
    sagaFetch.next({
        json: getJsonMock,
      }) // yield steps.json()
    expect(getJsonMock).toBeCalled()
    const put = sagaFetch.next(stepsObj) // every time you call next you have to pass object or it will be undefined 
    expect(put.value.type).toEqual('PUT')
    expect(JSON.stringify(put.value.payload.action)).toEqual(JSON.stringify(steps.setSteps(stepsObj)))
    expect(sagaFetch.next().done).toBeTruthy()
})


test('sagaFetch* with plan!',()=>{
    fetch.mockResponse(JSON.stringify(stepsObj))
    return expectSaga(steps.sagaFetch, 1,1)
    .put(steps.setSteps(stepsObj))
    .run()
})

test('sagaInterval* ',()=>{
    const saga = steps.sagaInterval()
    expect(saga.next().value.type).toEqual('SELECT')
    expect(saga.next().value.type).toEqual('SELECT')
    expect(saga.next().value.type).toEqual('FORK') // test it like that
    const delayb = saga.next()
    expect(delayb.value.type).toEqual('CALL')
    expect(delayb.value).toEqual(delay(10000)) // or liker that! 
   // infinite loop
})