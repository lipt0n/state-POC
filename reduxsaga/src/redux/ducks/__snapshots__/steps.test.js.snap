// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeStepConfig action creator 1`] = `
Object {
  "payload": Object {
    "config": Object {
      "increment": 0,
    },
    "stepId": 1,
  },
  "type": "STEPS:CHANGE",
}
`;

exports[`getSteps action creator 1`] = `
Object {
  "type": "STEPS:GET",
}
`;

exports[`runStep action creator 1`] = `
Object {
  "payload": Object {
    "stepId": 1,
  },
  "type": "STEPS:RUN",
}
`;

exports[`setSteps action creator 1`] = `
Object {
  "payload": Array [
    Object {
      "configuration": Array [
        Object {
          "value": 1,
        },
      ],
      "id": 0,
      "inputs": Array [],
      "last_executed_at": "2019-03-25T21:50:58.643Z",
      "operator_id": "load_int",
      "outputs": Array [
        Object {
          "value": 1,
        },
      ],
      "status": "up_to_date",
    },
    Object {
      "configuration": Array [
        Object {
          "increment": 0,
        },
      ],
      "id": 1,
      "inputs": Array [
        Object {
          "other_step_id": 0,
          "other_step_type_output_id": "value",
          "step_type_input_id": "value",
        },
      ],
      "last_executed_at": "2019-03-25T22:00:41.906Z",
      "operator_id": "increment_int",
      "outputs": Array [
        Object {
          "value": 1,
        },
      ],
      "status": "up_to_date",
    },
    Object {
      "configuration": Array [
        Object {
          "increment": 1,
        },
      ],
      "id": 2,
      "inputs": Array [
        Object {
          "other_step_id": 1,
          "other_step_type_output_id": "value",
          "step_type_input_id": "value",
        },
      ],
      "last_executed_at": "2019-03-25T22:01:11.145Z",
      "operator_id": "increment_int",
      "outputs": Array [
        Object {
          "value": 2,
        },
      ],
      "status": "up_to_date",
    },
    Object {
      "configuration": Array [
        Object {
          "increment": 0,
        },
      ],
      "id": 3,
      "inputs": Array [
        Object {
          "other_step_id": 2,
          "other_step_type_output_id": "value",
          "step_type_input_id": "value",
        },
      ],
      "last_executed_at": "2019-03-25T22:01:13.904Z",
      "operator_id": "increment_int",
      "outputs": Array [
        Object {
          "value": 2,
        },
      ],
      "status": "up_to_date",
    },
  ],
  "type": "STEPS:SET",
}
`;

exports[`stepsReducer 1`] = `
Array [
  Object {
    "configuration": Array [
      Object {
        "value": 1,
      },
    ],
    "id": 0,
    "inputs": Array [],
    "last_executed_at": "2019-03-25T21:50:58.643Z",
    "operator_id": "load_int",
    "outputs": Array [
      Object {
        "value": 1,
      },
    ],
    "status": "up_to_date",
  },
  Object {
    "configuration": Array [
      Object {
        "increment": 0,
      },
    ],
    "id": 1,
    "inputs": Array [
      Object {
        "other_step_id": 0,
        "other_step_type_output_id": "value",
        "step_type_input_id": "value",
      },
    ],
    "last_executed_at": "2019-03-25T22:00:41.906Z",
    "operator_id": "increment_int",
    "outputs": Array [
      Object {
        "value": 1,
      },
    ],
    "status": "up_to_date",
  },
  Object {
    "configuration": Array [
      Object {
        "increment": 1,
      },
    ],
    "id": 2,
    "inputs": Array [
      Object {
        "other_step_id": 1,
        "other_step_type_output_id": "value",
        "step_type_input_id": "value",
      },
    ],
    "last_executed_at": "2019-03-25T22:01:11.145Z",
    "operator_id": "increment_int",
    "outputs": Array [
      Object {
        "value": 2,
      },
    ],
    "status": "up_to_date",
  },
  Object {
    "configuration": Array [
      Object {
        "increment": 0,
      },
    ],
    "id": 3,
    "inputs": Array [
      Object {
        "other_step_id": 2,
        "other_step_type_output_id": "value",
        "step_type_input_id": "value",
      },
    ],
    "last_executed_at": "2019-03-25T22:01:13.904Z",
    "operator_id": "increment_int",
    "outputs": Array [
      Object {
        "value": 2,
      },
    ],
    "status": "up_to_date",
  },
]
`;
